File: algorithm/mergesort/mergesort.py
def merge_sort(li):
    if len(li)<2:
        return li
    
    m=len(li)/2
    r=li[:m]
    l=li[m:]
    return merge(merge_sort(r), merge_sort(l))  

def merge(l, r):
    i=0
    j=0
    retList=[]
    while i<len(l) and j<len(r):
        if(l[i]<r[j]):
            retList.append(l[i])
            i+=1
        else:
            retList.append(r[j])
            j+=1
    retList+=l[i:]
    retList+=r[j:]
    return retList


def main():
    li=[1,2,5,3,8,4,7]
    #print merge_sort(li);
    print merge_sort(li) 

if __name__=="__main__":
    main()

File: algorithm/mergesort/mergesort.java

import java.util.*;

public class mergesort{
	public static void main(String[] args) {
		List<Integer> r=new ArrayList<Integer>();
		List<Integer> l=new ArrayList<Integer>();
		List<Integer> list1=new ArrayList<Integer>();

		int[] x={1,2,5,7,9};
		int[] y={3,4,8,10,20};
		int[] list={3,1,4,2,8,10,9,7,20};

		createList(l,x);
		createList(r,y);
		createList(list1,list);


		System.out.println(merge_sort(list1).toString());
		//appendList(3,r,l);
		//System.out.println(r.toString());
		

	}

	public static List merge_sort(List<Integer> list1){
		if(list1.size()<2) return list1;
		int m=list1.size()/2;
		List<Integer> l=ListRange(list1, 0, m);
		List<Integer> r=ListRange(list1, m, list1.size());
		return merge(merge_sort(r), merge_sort(l));
	}

	public static List<Integer> merge(List<Integer> l, List<Integer> r){
		int i,j;
		i=j=0;
		List<Integer> retList=new ArrayList<Integer>();
		while(i<l.size() && j<r.size()){
			if(l.get(i)<r.get(j)){
				retList.add(l.get(i));
				i++;
			}
			else{
				retList.add(r.get(j));
				j++;
			}
		}
		appendList(i, retList,l);
		appendList(j, retList,r);
		return retList;

	}


	public static void createList(List<Integer> l, int[] arr){
		for(int ele:arr){
			l.add(ele);
		}
	}

	public static void appendList(int idx, List<Integer> l,List<Integer> r){
		if(idx>=r.size()) return;
		for(int i=idx; i<r.size();i++){
			l.add(r.get(i));
		}

	}

	public static List<Integer> ListRange(List<Integer> list1, int start, int end){
		List<Integer> retList=new ArrayList<Integer>();
		for(int i=start; i<end; i++)
		{
			retList.add(list1.get(i));
		}
		return retList;

	}

}

